Graphics engine

jefbelmans opened the pull request:
# Basic graphics engine foundation
Added basic mesh rendering capabilities to the graphics engine. Can now render multiple objects with a simple PosNormCol shaders (Normal is used for the color).
## Important changes
**MOST IMPORTANT:** in order to build the code, the [Vulkan SDK ](https://vulkan.lunarg.com/#new_tab )must be installed!
The Vulkan SDK includes and libs have been removed from the 3rdParty folder and are now added via the environment variables using CMakes "find_package()" method.

Needed to add a few libraries to the main project and cleaned up the CMake project a bit:
- Moved all third party libraries into its own CMakeLists inside the "3rdParty" directory
- Added vkbootstrap lib
- Added Vulkan Memory Allocator lib
- Added TinyObjLoader lib

jefbelmans committed:
Created command pool & buffer

Added a command pool and a single buffer of the graphics queue family. This buffer will be used to record the graphics commands

jefbelmans committed:
Added a default render pass

Added a default render pass that will be used to write to the images using render commands.

jefbelmans committed:
Added a basic render loop

Added a basic render loop that flashes the clear color green.

jefbelmans committed:
Added basic vert and frag shader

Added a very basic vertex and fragment shader.

Added Vulkan shader compilation to project. Compiled shaders will now be output to the /Shaders/Compiled folder in both the project and the build folder.

jefbelmans committed:
Loaded the shaders into the graphics pipeline

jefbelmans committed:
Finished the graphics pipeline

Added a helper class that builds a graphics pipeline. Used it to render a triangle to the viewport.

jefbelmans committed:
Added deletion queue

Added a deletion queue that neatly cleans up all the Vulkan objects.

jefbelmans committed:
Removed Vulkan files from project and added VMA

Removed the Vulkan files and libraries from the project as they are now added via the SDK install directory.

Added the Vulkan Memory Allocator library from AMD. Added the includes to the root CMakeLists.

jefbelmans committed:
Cleaned up project

Moved all third party includes to a separate CMakeLists inside the 3rdParty directory.

Moved the VkBootstrap library from the Graphics Engine to the 3rdParty directory.

jefbelmans committed:
Added basic Vertex and Mesh struct

Added a basic Vertex and Mesh struct to represent a mesh. Used in Pipeline Layout and Push Constants.

Added a basic rotating triangle using push constants to rotate it.

jefbelmans committed:
Added tinyobjloader and Utah teapot

Added tinyobjloader library to load obj files.

Loaded and added the Utah teapot.

jefbelmans committed:
Added depth testing

Added depth testing to the render pipeline. Added all the required initializers.

jefbelmans committed:
Added basic scene and object management

Added a way to more easily draw and upload multiple meshes and cleaned up the code base a bit.

LeeVangraefschepe commented on Graphics/VulkanEngine.h:
```diff
@@ -30,14 +72,70 @@ namespace leap::graphics
+		VmaAllocator m_Allocator{ VK_NULL_HANDLE };
+
+		// Mesh
+		Mesh m_TeapotMesh;
```
Temp?

jefbelmans replied:
Is indeed temp and only used during development of the core features. This helps me rappidly test and diagnose new feature implementations. Should get removed and reworked into the main engine once everything is setup.

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -46,27 +57,133 @@ void VulkanEngine::Initialize()
 {
-	
+	// Wait for the GPU to be done with the last frame, 1s timeout
+	VK_CHECK(vkWaitForFences(m_Device, 1, &m_RenderFence, true, 1000000000));
```
Replace with vkGetFenceStatus
Don't hardcode timeouts

jefbelmans replied:
Tested it with 10k objects with a total of 22.5mil triangles, double buffering and benchemarked it: there is no noticable improvement in frametimes.
Samsung recommends against waiting for fences on their mobile devices (and probably other mobile devices in general) since the CPU might be waiting a noticable amount of time for the GPU to finish rendering. While the CPU is waiting, it could then execute small jobs that take less than 1ms, as to not make the GPU wait whenever it is ready.
vkWaitForFences offers a lot of flexibility and is well optimised by the drivers and API. I will, however, replace the hardcoded timeout by UINT64_MAX.

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -46,27 +57,133 @@ void VulkanEngine::Initialize()
+
+	// Request image from the swap chain, wait 1 second if none available
+	uint32_t swapchainImageIndex;
+	VK_CHECK(vkAcquireNextImageKHR(m_Device, m_SwapChain, 1000000000, m_PresentSemaphore, nullptr, &swapchainImageIndex));
```
Same here

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -46,27 +57,133 @@ void VulkanEngine::Initialize()
+	rpInfo.clearValueCount = 2;
+	rpInfo.pClearValues = &clearValues[0];
+
+	// Begin the render pass
```
Rebegin? Begin was also noted on L107

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -145,4 +275,497 @@ void VulkanEngine::InitializeSwapChain()
+{
+	glm::vec3 camPos{0.f, -10.f, -300.f};
+
+	// WVP matrix
```
Not WVP but VP matrix

jefbelmans replied:
WVP matrix is defined at line 727

LeeVangraefschepe reviewed (changes requested):
Check comments

jefbelmans commented:
There's indeed still a lot of hardcoded and temp values, these are going to get reworked once everyrhing is setup and works.

jefbelmans committed:
Added descriptor sets

Added descriptor sets with support for static and dynamic uniform buffers.

Renamed the main shader and removed the temp ones.

jefbelmans committed:
Added SSBO support

LeeVangraefschepe commented on Graphics/CMakeLists.txt:
```diff
@@ -1,4 +1,51 @@
-#Graphics engine cmake
-add_library(GraphicsEngine STATIC Renderer.h Renderer.cpp VulkanTypes.h VulkanInitializers.h VulkanInitializers.cpp VulkanEngine.h VulkanEngine.cpp VkBootstrap.h VkBootstrap.cpp)
+# Graphics engine
+add_library(GraphicsEngine STATIC Renderer.h Renderer.cpp VulkanTypes.h VulkanInitializers.h VulkanInitializers.cpp VulkanEngine.h VulkanEngine.cpp VulkanPipelineBuilder.h VulkanPipelineBuilder.cpp "vk_mesh.h" "vk_mesh.cpp")
```
Why the strings?

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -145,4 +283,725 @@ void VulkanEngine::InitializeSwapChain()
+	
+	for (int i = 0; i < MAX_FRAMES_IN_FLIGHT; i++)
+	{
+		VK_CHECK(vkCreateCommandPool(m_Device, &commandPoolInfo, nullptr, &m_Frames[i].commandPool));
```
Cache m_Frames[i]

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -145,4 +283,725 @@ void VulkanEngine::InitializeSwapChain()
+	const size_t sceneDataBufferSize = MAX_FRAMES_IN_FLIGHT * PadUniformBufferSize(sizeof(GPUSceneData));
+	m_SceneDataBuffer = CreateBuffer(sceneDataBufferSize, VK_BUFFER_USAGE_UNIFORM_BUFFER_BIT, VMA_MEMORY_USAGE_CPU_TO_GPU);
+	
+	const int MAX_OBJECTS = 10'000;
```
Hardcoded

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -145,4 +283,725 @@ void VulkanEngine::InitializeSwapChain()
+	
+	for (int i = 0; i < MAX_FRAMES_IN_FLIGHT; i++)
+	{
+		VK_CHECK(vkCreateFence(m_Device, &fenceCreateInfo, nullptr, &m_Frames[i].renderFence));
```
Again cache m_Frames[i] or use ranged based for loop

LeeVangraefschepe commented on Graphics/VulkanEngine.h:
```diff
@@ -1,10 +1,90 @@
+	};
+
+	// Number of frames that can be processed concurrently
+	constexpr uint8_t MAX_FRAMES_IN_FLIGHT = 2;
```
Global var

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -145,4 +283,725 @@ void VulkanEngine::InitializeSwapChain()
+	file.seekg(0);
+
+	// Load the entire file into the buffer
+	file.read(reinterpret_cast<char*>(buffer.data()), fileSize);
```
signed unsigned conversion

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -145,4 +283,725 @@ void VulkanEngine::InitializeSwapChain()
+
+Material* VulkanEngine::CreateMaterial(const std::string& name, VkPipeline pipeline, VkPipelineLayout layout)
+{
+	Material mat;
```
Initialize

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -145,4 +283,725 @@ void VulkanEngine::InitializeSwapChain()
+	glm::mat4 proj{glm::perspective(glm::radians(70.f), (float)m_WindowExtent.width / (float)m_WindowExtent.height, 0.1f, 1000.f)};
+	proj[1][1] *= -1;
+
+	GPUCameraData camData;
```
Initialize

LeeVangraefschepe commented on Graphics/VulkanEngine.cpp:
```diff
@@ -145,4 +283,725 @@ void VulkanEngine::InitializeSwapChain()
+
+	Material* defaultMaterial = GetMaterial("defaultmesh");
+	Mesh* teapotMesh = GetMesh("Teapot");
+	const int vertCount = teapotMesh->vertices.size();
```
signed unsigned conversion

LeeVangraefschepe commented on Graphics/VulkanEngine.h:
```diff
@@ -1,10 +1,90 @@
+
 	class VulkanEngine final
 	{
 	public:
```
Rule of 5? Time to make a copy :p

LeeVangraefschepe commented on Graphics/VulkanInitializers.cpp:
```diff
@@ -1 +1,254 @@
+
+VkImageCreateInfo vkinit::ImageCreateInfo(VkFormat format, VkImageUsageFlags usageFlags, VkExtent3D extent)
+{
+	VkImageCreateInfo info = { };
```
= {} ?

LeeVangraefschepe commented on Graphics/VulkanInitializers.cpp:
```diff
@@ -1 +1,254 @@
+VkImageViewCreateInfo vkinit::ImageViewCreateInfo(VkFormat format, VkImage image, VkImageAspectFlags aspectFlags)
+{
+	//build a image-view for the depth image to use for rendering
+	VkImageViewCreateInfo info = {};
```
= {} ?

LeeVangraefschepe commented on Graphics/VulkanPipelineBuilder.h:
```diff
@@ -0,0 +1,23 @@
+{
+	class PipelineBuilder final
+	{
+	public:
```
Public m_variables?
Maybe struct?

LeeVangraefschepe commented on Graphics/vk_mesh.cpp:
```diff
@@ -0,0 +1,124 @@
+    if (!err.empty())
+    {
+        std::cerr << err << std::endl;
+        return false;
```
Don't return use throw

LeeVangraefschepe reviewed (changes requested):
Const correctness and Cmake on github isnt working
